Fix last month not being displayed if last date is selected in DateRangePicker

b1acKr0se opened the pull request:
## Description

The condition to set the `_initialMonthIndex` is missing the case if the `lastDate` is the same as `initialDate`, so if the `lastDate` is selected, then the last month will not be visible to the user at launch.

| Current | With this fix |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/8596158/94153617-30cdab00-fea7-11ea-9ceb-8b075c46d0e8.gif" width="200"> | <img src="https://user-images.githubusercontent.com/8596158/94153504-109dec00-fea7-11ea-8444-c7baf47ef256.gif" width="200"> |


## Related Issues

I've just found it.

## Tests

I added the following tests:

- A test which sets the `initialDateRange` to be the same as the `lastDate` supplied to the `DateRangePicker`. The last month should be visible in this case.
- A test which sets the `initialDateRange` to be the same as the `firstDate` supplied to the `DateRangePicker`. The first and part of the second month should be visible in this case. 
- A test which does not set the `initialDateRange`. The month from the given `currentDate` should be visible in this case.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

b1acKr0se committed:
Fix last month not being displayed if last date is selected in DateRangePicker

b1acKr0se committed:
Made tests clearer

darrenaustin commented on packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:
```diff
@@ -110,7 +110,8 @@ class _CalendarDateRangePickerState extends State<CalendarDateRangePicker> {
     final DateTime initialDate = widget.initialStartDate ?? widget.currentDate;
     if (widget.firstDate.isBefore(initialDate) &&
-        widget.lastDate.isAfter(initialDate)) {
+        widget.lastDate.isAfter(initialDate) ||
```
Nice catch. However, while we don't see the same problem with `firstDate` here, that is only because it `_initialMonthIndex` defaults to 0. I would still like to see both cases handled here so that it is obvious both are handled. Perhaps something like:

```dart
    if (!initialDate.isBefore(widget.firstDate) && 
        !initialDate.isAfter(widget.lastDate)) {
      _initialMonthIndex = utils.monthDelta(widget.firstDate, initialDate);
    }
```

b1acKr0se replied:
Oh this is much better! Thank you.

darrenaustin commented on packages/flutter/test/material/date_range_picker_test.dart:
```diff
@@ -122,6 +122,90 @@ void main() {
+    }
+
+    await preparePicker(tester, (Future<DateTimeRange> range) async {
+      final DateTime currentMonth = DateTime.now();
```
It is not a good practice to have tests dependent on things like the current date, as the results could vary between runs. A better practice here would be to just set the `currentDate` to a known constant and check against that. That way you wouldn't need to compute the month names as you would be testing against the one that you set.

b1acKr0se replied:
I was not aware that `DateRangePicker` allows `currentDate` to be set, haha. Thank you for the best practice advice!

darrenaustin commented on packages/flutter/test/material/date_range_picker_test.dart:
```diff
@@ -372,10 +456,10 @@ void main() {
         expect(find.text('January 2016'), findsOneWidget);
         expect(find.text('February 2016'), findsOneWidget);
-        expect(find.text('Mar 2016'), findsNothing);
+        expect(find.text('March 2016'), findsNothing);
```
Nice catch. Thanks.

darrenaustin reviewed (changes requested):
Wow, nice work. Thanks so much for catching this and contributing a fix for it (with tests included!). Really appreciate it. I have a few requests below, but otherwise it looks good. Thx!

b1acKr0se committed:
Made condition checking for initial index calculation more obvious and improve tests

b1acKr0se commented:
@darrenaustin Thank you for the review! I've addressed all of your comments, so please take a look when you have the time.

Also, looks like the `customer_testing-linux` CI failed because of a timeout. How do I trigger it to re-run the test?

darrenaustin reviewed (approved):
LGTM. Thanks so much for the contribution. Nice work!

darrenaustin commented:
> Also, looks like the `customer_testing-linux` CI failed because of a timeout. How do I trigger it to re-run the test?

You should be able to click on the "Details" link and there should be a re-run link next the job. Perhaps this only shows up for maintainers though? I went ahead and restarted it. I will keep an eye on this and make sure it completes. Thanks again.

b1acKr0se commented:
@darrenaustin Yeah I saw that button once I logged in to CirrusCI. Thank you so much for taking the time to review this!
